### PR TITLE
Simplify function send in reply.js

### PR DESCRIPTION
**Pull request by trivikr**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: lib/reply.js [hunk 1]
```diff
@@ -33,46 +33,31 @@ Reply.prototype.send = function (payload) {
...
-
-  const contentType = this.res.getHeader('Content-Type')
-  const contentTypeNotSet = contentType === undefined
-
-  if (payload !== null) {
-    if (Buffer.isBuffer(payload) || typeof payload.pipe === 'function') {
+  if (payload !== undefined) {
+    const contentType = this.res.getHeader('Content-Type')
+    const contentTypeNotSet = contentType === undefined
+    if (isPayloadStream(payload)) {
```
Comment (mcollina) [inline lib/reply.js]: can you restore this check inline? It increases our inline count for no particular reason.

File: lib/reply.js [hunk 2]
```diff
@@ -33,46 +33,31 @@ Reply.prototype.send = function (payload) {
...
       if (contentTypeNotSet) {
         this.res.setHeader('Content-Type', 'application/octet-stream')
       }
-      onSendHook(this, payload)
-      return
-    }
-
-    if (contentTypeNotSet && typeof payload === 'string') {
+    } else if (contentTypeNotSet && isPayloadStringType(payload)) {
```
Comment (mcollina) [inline lib/reply.js]: Same here

File: lib/reply.js (hunk 1 above)
Comment (trivikr) [reply lib/reply.js]: Done

File: lib/reply.js (hunk 2 above)
Comment (trivikr) [reply lib/reply.js]: Done

File: lib/reply.js [hunk 3]
```diff
@@ -33,46 +33,31 @@ Reply.prototype.send = function (payload) {
...
   }
-
   this.sent = true
 
   if (payload instanceof Error || this._isError === true) {
     handleError(this, payload, onSendHook)
     return
   }
 
-  if (payload === undefined) {
```
Comment (jsumners) [inline lib/reply.js]: Why is this short circuit removed?

Comment (trivikr) [reply lib/reply.js]: It's changed to:
```javascript
if (payload !== undefined)
```
So that `onSendHook(this, payload)` is called just once

**Comment by jsumners** [comment]
I'm 👎 on this. The sole purpose of this PR is to reduce the number of code occurrences of `onSendHook` within the function at the cost of eliminating short circuits. I'm also not entirely convinced the logic is the same due to the fact that the original code checks for a serializer _after_ determining the content type and this PR checks for the serializer in an `else if` along with  determining the content type.

It seems to me that tests may not be covering this method correctly (sufficiently).

fastify

**Comment by trivikr** [comment]
This PR doesn't perform any additional computation because of removal of those elimination of short circuits, so it won't affect performance.
And the short circuits are removed only where `onSendHook(this, payload)` is called, the short circuit for `this.sent` or `payload instanceof Error` aren't removed as they don't call `onSendHook` directly.

It can be seen in the screenshots provided by @jsumners above, that this PR improves readability of the function, so it'll be good for developers.

**Comment by jsumners** [comment]
@trivikr okay, in your version of the code: if `payload !== null` and `this._serializer` is set, are the same set of operations performed as would be performed on the `master` branch?

**Comment by trivikr** [comment]
@jsumners `payload !== null` and `this._serializer` is set:

In master branch:
* if `payload !== null`
  * if `(Buffer.isBuffer(payload) || typeof payload.pipe === 'function')` then `return`
  * if `(contentTypeNotSet && typeof payload === 'string')` then `return`
* if `(this._serializer)`
  
In PR:
* if `(payload !== null && (Buffer.isBuffer(payload) || typeof payload.pipe === 'function')`
* else if `(contentTypeNotSet && payload !== null && typeof payload === 'string')`
* else if `(this._serializer)`

I don't see any use case missing, and one additional `payload !== null` is performed in PR

**Comment by nwoltman** [comment]
I'm -1 on this. The code is correct, but it has redundant type checking. See here:

```js
if (payload !== null && (Buffer.isBuffer(payload) || typeof payload.pipe === 'function')) {
  // Handle Buffer or stream
} else if (contentTypeNotSet && typeof payload === 'string') {
  this.res.setHeader('Content-Type', 'text/plain')
}
```

If the `payload` is `null`, the first `if` will see this and move on to the `else if`, but the `else if` code is unnecessary because if the payload is `null` then the result of that `else if` will always be `false`. The original code avoids additional type checks once the type of the payload has been determined.

**Comment by jsumners** [comment]
> The original code avoids additional type checks once the type of the payload has been determined.

i.e. the short circuits are eliminated.

**Comment by jsumners** [comment]
@trivikr you're right. I was overlooking the `return` statements.

**Comment by trivikr** [comment]
Any update on this PR?

**Comment by mcollina** [comment]
@trivikr can you address @nwoltman comment https://github.com/fastify/fastify/pull/751#issuecomment-363136384?

**Comment by trivikr** [comment]
The redundant type-checking mentioned in https://github.com/fastify/fastify/pull/751#issuecomment-363136384 was added to make code simpler to read.
Closing the PR as it's pretty old, and there were multiple -1s on this

**Pull request closed without merging**